### PR TITLE
Add collisions and gravity

### PR DESCRIPTION
**Pull request by dm-gh**

### CODE REVIEW

File: src/Model/Map.hs [hunk 1]
```diff
@@ -28,35 +28,33 @@ initialMap :: Map
 initialMap = foldr (\index acc -> addMapRow index acc) [] [0..(levelRowNumber initialLevel - 1)]
 
 addMapRow :: Int -> Map -> Map
-addMapRow row list = foldr (\index acc -> if even index
-                                          then addBasicMapTile (Solid, floorTexture) row index acc
-                                          else addBasicMapTile (Transparent, backgroundTexture) row index acc)
+addMapRow row list = foldr (\col acc -> if row + col < 30
```
Comment (alois9866) [inline src/Model/Map.hs]: Change number 30 to some named constant.

File: src/Model/CommonTypes.hs [hunk 2]
```diff
@@ -141,12 +194,20 @@ type Map = [MapRow]
...
-data Tile = Tile
-  { tileType :: TileType
-  , tileObject :: Object
-  } deriving (Show)
+instance Eq Tile where
+  (==) (Solid _) (Transparent _) = False
+  (==) (Transparent _) (Solid _) = False
+  (==) (Solid a1) (Solid a2) = appearanceBox a1 == appearanceBox a2
+  (==) (Transparent a1) (Transparent a2) = appearanceBox a1 == appearanceBox a2
+
```
Comment (alois9866) [inline src/Model/CommonTypes.hs]: Add type signature.

File: src/Model/CommonTypes.hs [hunk 3]
```diff
@@ -141,12 +194,20 @@ type Map = [MapRow]
...
+isSolid (Solid _) = True
+isSolid (Transparent _) = False
```
Comment (alois9866) [inline src/Model/CommonTypes.hs]: Change `(Tranparent _)` to just `_` for future extensibility.

File: src/Model/CommonTypes.hs [hunk 4]
```diff
@@ -49,6 +63,39 @@ mulByNumber (Vector (x, y)) number = Vector (x * number, y * number)
...
+dotProduct :: Vector -> Vector -> Float
+dotProduct (Vector (x1, y1)) (Vector (x2, y2)) = x1 * x2 + y1 * y2
+
+angleCosBetweenVectors :: Vector -> Vector -> Float
+angleCosBetweenVectors v1 v2 = (v1 `dotProduct` v2) / (vectorLength v1 * vectorLength v2)
+
+perpendicularVector :: Vector -> Vector
+perpendicularVector (Vector (x, y)) = Vector (y, -x)
+
+defineDescartesFourth :: Vector -> Int
```
[the remark this answers is not in the document]
Comment (dm-gh) [reply src/Model/CommonTypes.hs]: done

File: src/Model/Map.hs (hunk 1 above)
Comment (dm-gh) [reply src/Model/Map.hs]: why? this function does not  require further extension and this constant is not used anywhere else.

File: src/Model/CommonTypes.hs (hunk 2 above)
Comment (dm-gh) [reply src/Model/CommonTypes.hs]: why?

File: src/Model/CommonTypes.hs (hunk 3 above)
Comment (dm-gh) [reply src/Model/CommonTypes.hs]: done

File: src/Model/CommonTypes.hs (hunk 2 above)
Comment (dm-gh) [reply src/Model/CommonTypes.hs]: done

**Pull request merged**